neural_cherche/retrieve/bm25.py

```diff
@@ -0,0 +1,203 @@
+    def add(
+        self,
+        documents_embeddings: dict[str, csr_matrix],
+    ) -> "CountVectorizer":
+        """Add new documents to the CountVectorizer retriever."""
+        for document_key in documents_embeddings:
+            self.documents.append({self.key: document_key})
+
+        self.n_documents += len(documents_embeddings)
+        n_samples,n_features  = self.matrix.shape
+        df = np.bincount(self.matrix.indices, minlength=n_features)#Count the number of non-zero values for each feature in sparse X
+        idf = np.log(n_samples / (df)) #compute the log of idf
+        avdl  = self.matrix.sum(1).mean() # mean of all 
+        len_X = self.matrix.sum(1).A1 # the length of each doc
+        self.denom =  (self.k1 * (1 - self.b + self.b * len_X / avdl))
+        self.numer = self.matrix.multiply(np.broadcast_to(idf, (n_samples,n_features) )) 
+
+        return self
+    def encode_queries(self, queries: list[dict]) -> dict[str, csr_matrix]:
+        """Encode queries into sparse matrix.
+
+        Parameters
+        ----------
+        queries
+            Queries to encode.
+
+        """
+        if not self.fit:
+            raise ValueError("You must call the `encode_documents` method first.")
+
+        # matrix is a csr matrix of shape (n_queries, n_features)
+        content = [
+                " ".join([doc.get(field, "") for field in self.on]) for doc in queries
+            ]
+        matrix = self.countVector.transform(raw_documents=content)
+
+        return {query[self.key]: row for query, row in zip(queries, matrix)}
+    def __call__(
+        self,
+        queries_embeddings: dict[str, csr_matrix],
+        k: int = None,
+        batch_size: int = 2000,
+        tqdm_bar: bool = True,
+    ) -> list[list[dict]]:
+        """Retrieve documents from batch of queries.
+
+        Parameters
+        ----------
+        queries_embeddings
+            Queries embeddings.
+        k
+            Number of documents to retrieve. Default is `None`, i.e all documents that match the
+            query will be retrieved.
+        batch_size
+            Batch size to use to retrieve documents.
+        """
+        k = k if k is not None else self.n_documents
+
+        ranked = []
+        count_batch=0
+        for batch_queries in utils.batchify(
+            list(queries_embeddings.values()),
+            batch_size=batch_size,
+            desc=f"{self.__class__.__name__} retriever",
+            tqdm_bar=tqdm_bar,
+        ):
+            
+            batch_match, batch_similarities = self.top_k(batch_queries=batch_queries, k=k)
+
+            for match, similarities in zip(batch_match, batch_similarities):
+                
+                ranked.append(
+                    [
+                        {**self.documents[idx], "similarity": similarity}
+                        for idx, similarity in zip(match, similarities)
+                        #if similarity > 0
```

Comment:
We did filter similarity > 0 in order to only retrieve top_k documents that match the query. Does the function top_k only return documents with similarity > 0 ? Otherwise I think we should use this condition

Reply:
I was testing something I will uncomment it back